iothub_client/samples/pnp/pnp_temperature_controller/pnp_thermostat_component.c

```diff
@@ -0,0 +1,367 @@
+// Format string for sending maxTempSinceLastReboot property
+static const char g_maxTempSinceLastRebootPropertyFormat[] = "%.2f";
+// Format of the body when responding to a targetTemperature 
+static const char g_targetTemperaturePropertyResponseFormat[] = "%.2f";
+
+// Start time of the program, stored in ISO8601 format string for UTC
+char g_programStartTime[128] = {0};
+
+
+// Response description is an optional, human readable message including more information
+// about the setting of the temperature.  Because we accept all temperature requests, we 
+// always indicate a success.  An actual sensor could optionally return information about
+// a temperature being out of range or a mechanical issue on the device on error cases.
+static const char g_temperaturePropertyResponseDescription[] = "success";
+
+//
+// PNP_THERMOSTAT_COMPONENT simulates a thermostat component
+// (as in thermostat1 or thermostat2 in the TemperatureController model).  We need separate data structures
+// because the components can be independently controlled.
+//
+typedef struct PNP_THERMOSTAT_COMPONENT_TAG
+{
+    // Name of this component
+    char componentName[PNP_MAXIMUM_COMPONENT_LENGTH + 1];
+    // Current temperature of this thermostat component
+    double currentTemperature;
+    // Minimum temperature this thermostat has been at during current execution run of this thermostat component
+    double minTemperature;
+    // Maximum temperature thermostat has been at during current execution run of this thermostat component
+    double maxTemperature;
+    // Number of times temperature has been updated, counting the initial setting as 1.  Used to determine average temperature of this thermostat component
+    int numTemperatureUpdates;
+    // Total of all temperature updates during current exceution run.  Used to determine average temperature of this thermostat component
+    double allTemperatures;
+}
+PNP_THERMOSTAT_COMPONENT;
+
+//
+// BuildUtcTimeFromCurrentTime writes the current time, in ISO 8601 format, into the specified buffer
+//
+static bool BuildUtcTimeFromCurrentTime(char* utcTimeBuffer, size_t utcTimeBufferSize)
+{
+    bool result;
+    time_t currentTime;
+    struct tm * currentTimeTm;
+
+    time(&currentTime);
+    currentTimeTm = gmtime(&currentTime);
+
+    if (snprintf(utcTimeBuffer, utcTimeBufferSize, g_ISO8601Format, currentTimeTm->tm_year + 1900, currentTimeTm->tm_mon, currentTimeTm->tm_mday, 
+                  currentTimeTm->tm_hour, currentTimeTm->tm_min, currentTimeTm->tm_sec) < 0)
+    {
+        LogError("snprintf on UTC time failed");
+        result = false;
+    }
+    else
+    {
+        result = true;
+    }
+
+    return result;
+}
+
+
+PNP_THERMOSTAT_COMPONENT_HANDLE PnP_ThermostatComponent_CreateHandle(const char* componentName)
+{
+    PNP_THERMOSTAT_COMPONENT* thermostatComponent;
+
+    if (strlen(componentName) > PNP_MAXIMUM_COMPONENT_LENGTH)
```

Comment:
I suppose I don't have to, but if I don't make this check then I end up over-running the buffer if a length is too long.  I get it's sample code but having spent a career in OS's won't let me sleep well if I do that.

Reply:
OK thought more on this and while it's not technically needed, the risk of someone taking this code as-is and buffer overrunning really, really would not let me sleep at night.,  A buffer overrun wouldn't necessarily crash and even Valgrind wouldn't be able to detect it if it didn't over-run `sizeof(PNP_THERMOSTAT_COMPONENT)`.  I've also considered just relying on the Hub to barf because the componentLength is too big other stuff would break, but if the hub changed that policy then we couldn't count on that.

Bottom line here this is a memory bounds check, not a policy check I can rely on Cloud for.  Only way to safely remove this would be to `malloc` the `componentName` field but I don't want to do that either.

Reply:
got you, but then ... should we check for propertyNames/commandNames length? whatever we choose, we should try to be consistent with other samples. I've not seen those checks in the comp-less sample.